README.md
# FWApplication

[![Pod Version](https://img.shields.io/cocoapods/v/FWApplication?style=flat)](http://cocoadocs.org/docsets/FWApplication/)
[![Pod Platform](https://img.shields.io/cocoapods/p/FWApplication.svg?style=flat)](http://cocoadocs.org/docsets/FWApplication/)
[![Pod License](https://img.shields.io/cocoapods/l/FWApplication.svg?style=flat)](https://github.com/lszzy/FWApplication/blob/master/LICENSE)
[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/lszzy/FWApplication)

# [FWFramework](https://github.com/lszzy/FWFramework)

# [中文](README_CN.md)

## Tutorial
iOS application framework, convenient for iOS development, compatible with OC and Swift.

All Swizzles in this framework will not take effect by default and will not affect existing projects. They need to be manually opened or invoked to take effect. This library has been used in formal projects, and will continue to be maintained and expanded in the future. Everyone is welcome to use and provide valuable comments to grow together.

## Installation
It is recommended to use CocoaPods or Swift Package Manager to install and automatically manage dependencies. For manual import, please refer to Example project configuration.

### CocoaPods
This framework supports CocoaPods, Podfile example:

	platform :ios, '11.0'
	use_frameworks!

	target 'Example' do
	  # Import the default subspecs
	  pod 'FWApplication'
	  
	  # Import the specified subspecs, see the podspec file for the list of subspecs
	  # pod 'FWApplication', :subspecs => ['FWApplication', 'Compatible', 'SDWebImage', 'Lottie']
	end

### Swift Package Manager
This framework supports Swift Package Manager, just add and check the required modules, Package example:

	https://github.com/lszzy/FWApplication.git
	
	# Check and import the default submodule
	import FWApplication
	
	# Check and import the specified sub-modules, see the Package.swift file for the list of sub-modules
	import FWApplicationCompatible
	import FWApplicationSDWebImage
	import FWApplicationLottie

## [Api](https://fwapplication.wuyong.site)
The document is located in the docs folder, just open index.html in the browser, or run docs.sh to automatically generate the Api document.

## [Changelog](CHANGELOG.md)
As this framework is constantly upgrading, optimizing and expanding new functions, the Api of each version may be slightly changed. If a compilation error is reported when the new version is upgraded, the solution is as follows:

	1. Just change to specify the pod version number to import, the recommended way, does not affect the project progress, upgrade to the new version only when you have time, example: pod'FWApplication', '3.8.1'
	2. Upgrade to the new version, please pay attention to the version update log. Obsolete Api will be migrated to the Deprecated submodule as appropriate, and will be deleted in subsequent versions

## Vendor
This framework uses a lot of third-party libraries. Thanks to the authors of all third-party libraries. I will not list them all here. For details, please refer to the relevant links of the source file.
 
	In the introduction of third-party libraries, in order to be compatible with existing project pod dependencies, as well as to customize changes and bug fixes of third-party libraries, and to facilitate subsequent maintenance, this framework uniformly modified the FW class prefix and fw method prefix. If there is any inconvenience during use, Please understand.
	If you are the author of a third-party open source library, if this library violates your rights, please let me know, and I will immediately remove the use of the third-party open source library. 

## Support
[wuyong.site](http://www.wuyong.site)


CHANGELOG.md
# Changelog

## [3.8.1] - 2022-08-03

### Changed
* Optimize PluginView progress bar view
* Optimize ViewContext to support subscribing to object and userInfo
* Optimize SwiftUI module directory structure

## [3.8.0] - 2022-08-02

### Added
* Added a large number of SwiftUI basic components to facilitate development
* Added methods to monitor View size and scroll view contentOffset
* Added a method to get the controller where View is located
* Added Introspect component and pull-down refresh component
* Added View configuration navigation bar method, which is common to UIKit
* Added PluginView for loading view specification plugins, common to UIKit
* Added HostingView for loading View as UIView
* Added StateView state machine view
* Added HostingController controller, common to UIKit
* Compatible with FWFramework 3.8.0 version

## [3.7.0] - 2022-07-27

### Added
* Added SwiftUI web image, wrapper, toolkit components
* Compatible with FWFramework 3.7.0 version

## [3.6.3] - 2022-07-18

### Added
* Added optional Lottie plugin to load Lottie animations
* Pull-down refresh plugin supports progress bar effect when dragging
* Pull-up additional plug-in support shows that there are no more views
* The shortcut jump TabBar is compatible with the case where the root controller is the navigation controller

## [3.6.2] - 2022-07-14

### Added
* FWTagCollectionView added selected font configuration

## [3.6.1] - 2022-07-13

### Changed
* Toast plugin supports controller configuration displayed in the window

## [3.6.0] - 2022-07-11

### Changed
* Refactor the image plugin to create a view method
* Modify the WebView bridge parameter context to weak reference
* Compatible with FWFramework 3.6.0 version

## [3.5.1] - 2022-07-06

### Fixed
* Image plugin supports ignore cache option
* Fix the problem that the empty interface occasionally displays abnormally
* Fix Tabbar controller loading remote image issue

## [3.5.0] - 2022-07-04

### Added
* Added batch registration WebView bridge class method
* Migrate some common methods to FWFramework
* Compatible with FWFramework 3.5.0 version

## [3.4.0] - 2022-06-26

### Changed
* Refactored OC version classification API, changed to fw_ prefix, removed Wrapper
* Added a new closing method for the pop-up plug-in
* Compatible with FWFramework 3.4.0 version

## [3.3.2] - 2022-06-20

### Changed
* Split popup plugin showAlert and showSheet methods
* Added custom style configuration parameters for popup plugins

## [3.3.1] - 2022-06-15

### Changed
* VC animation adds edge gesture close function
* Optimize the calculation problem in the background of the countDown method
* Fix the abnormal problem of navigation bar transition animation
* Compatible with FWFramework 3.3.1 version

## [3.3.0] - 2022-06-11

### Added
* New application framework configuration default template class
* Compatible with FWFramework 3.3.0 version

## [3.2.2] - 2022-06-08

### Changed
* Modify network component attribute names and parameter default values
* The toast pop-up window is not displayed when the optimization message is empty

## [3.2.1] - 2022-06-02

### Added
* Added Swift Network Mock component
* Added retry function when network request fails

## [3.2.0] - 2022-05-27

### Changed
* Refactor templateHeight to generic cache
* Optimize the TableViewDelegate proxy class
* Compatible with FWFramework 3.2.0 version

## [3.1.0] - 2022-05-26

### Added
* Refactor Swift version Api method
* Added FW global method, you can customize the call name
* Refactored some components and added several functions
* This version is not compatible with the previous version, the code must be migrated
* Compatible with FWFramework 3.1.0 version

## [3.0.0] - 2022-04-29

### Added
* Brand new version, using .fw. calling method
* Customizable .fw. for any call name
* Refactored some components and added several functions
* This version is not compatible with the previous version, the code must be migrated
* Compatible with FWFramework 3.0.0 version

## [1.4.0] - 2022-02-12

### Added
* FWViewController adds renderLayout hook method

### Changed
* FWToastPlugin message toast support does not automatically hide the method

## [1.3.1] - 2022-01-17

### Added
* Support controller or view to set plugins used individually

## [1.3.0] - 2022-01-15

### Added
* FWAlertPlugin supports custom style configuration
* Support FWApplication sub-module to be imported separately

## [1.2.1] - 2022-01-11

### Fixed
* Fix the problem that Swift cannot access the appearance singleton method

## [1.2.0] - 2021-12-31

### Added
* Refactored Pod sub-module, split OC and Swift code
* Support Swift Package Manager
* Added FWStatisticalObject to determine exposure and completion attributes

## [1.1.1] - 2021-12-30

### Changed

* Refactored FWStatisticalManager click and exposure buried point statistics implementation plan
* Unified Changed burying point is Click burying point statistics
* Support custom click and exposure buried point events, support exposure time

## [1.1.0] - 2021-12-28

### Added

* UICollectionViewFlowLayout supports section background settings
* Added FWCollectionViewAlignLayout component

## [1.0.4] - 2021-12-27

### Added

* Added FWCollectionViewAlignLayout component

## [1.0.3] - 2021-12-20

### Added

* Picture browser supports local picture path
* Added spacing configuration for custom pop-up plugins

## [1.0.2] - 2021-12-16

### Added

* The picture picker adds a fixed number of picture columns parameter

## [1.0.1] - 2021-12-15

### Added

* Plug-ins such as FWAlertPlugin support multi-level present
* The picture selector supports left and right scrolling during single selection preview

## [1.0.0] - 2021-12-09

### Added

* Split FWApplication from FWFramework, version 1.0.0 is released


FWApplication/Classes/Model/FWModel.h
/**
 @header     FWModel.h
 @indexgroup FWApplication
      FWModel
 @author     wuyong
 @copyright  Copyright © 2018 wuyong.site. All rights reserved.
 @updated    2018/9/26
 */

#import <Foundation/Foundation.h>
#import <objc/runtime.h>
#import "Foundation+FWApplication.h"

NS_ASSUME_NONNULL_BEGIN

/// 定义NSCopying实现宏
#define FWDefCopying() \
    - (id)copyWithZone:(NSZone *)zone \
    { \
        return [self fwModelCopy]; \
    }

/// 定义NSCoding实现宏
#define FWDefCoding( ) \
    - (instancetype)initWithCoder:(NSCoder *)aDecoder \
    { \
        return [self fwModelInitWithCoder:aDecoder]; \
    } \
    - (void)encodeWithCoder:(NSCoder *)aCoder \
    { \
        [self fwModelEncodeWithCoder:aCoder]; \
    }

/// 定义数组类型模型
#define FWModelArray( class ) \
    @protocol class <NSObject> \
    @end

/**
 *  Model转换协议
 *  数组类映射支持两种方式：
 *  1. 实现fwModelClassMapper方法，返回类映射字典。示例：@{@"books" : [Book class]}
 *  2. 声明Model类同名协议，同时定义数组属性时也声明协议。示例：NSArray<Book> *books
 *  Swift数组类映射时，需返回AnyClass类型。示例：["books": Book.self]
 */
@protocol FWModel <NSObject>

@optional

/// 属性映射，示例：@{@"name" : @"book_name", @"bookId" : [@"book_id", @"book.id"]}
+ (nullable NSDictionary<NSString *, id> *)fwModelPropertyMapper;

/// 类映射(Swift需使用AnyClass类型)，示例：@{@"books" : [Book class], @"users" : @"User"}
+ (nullable NSDictionary<NSString *, id> *)fwModelClassMapper;

/// 自定义字典解析类(Swift需使用AnyClass类型)
+ (nullable Class)fwModelClassForDictionary:(NSDictionary *)dictionary;

/// 属性黑名单列表
+ (nullable NSArray<NSString *> *)fwModelPropertyBlacklist;

/// 属性白名单列表
+ (nullable NSArray<NSString *> *)fwModelPropertyWhitelist;

/// 字典将要转换模型时钩子处理
- (NSDictionary *)fwModelWillTransformFromDictionary:(NSDictionary *)dictionary;

/// 字典转换模型时钩子处理
- (BOOL)fwModelTransformFromDictionary:(NSDictionary *)dictionary;

/// 模型转换字典时钩子处理
- (BOOL)fwModelTransformToDictionary:(NSMutableDictionary *)dictionary;

@end

/**
 *  ViewModel架构协议
 */
@protocol FWViewModel <NSObject>

@end

/**
 Model模型解析分类，参考自YYModel
 
 @see https://github.com/ibireme/YYModel
 */
@interface NSObject (FWModel)

/**
 从json创建对象，线程安全。NSDate会按照UTC时间解析，下同
 
 @param json json对象，支持NSDictionary、NSString、NSData
 @return 实例对象，失败为nil
 */
+ (nullable instancetype)fwModelWithJson:(id)json;

/**
 从字典创建对象，线程安全
 
 @param dictionary 字典数据
 @return 实例对象，失败为nil
 */
+ (nullable instancetype)fwModelWithDictionary:(NSDictionary *)dictionary;

/**
 从json创建Model数组
 
 @param json json对象，支持NSDictionary、NSString、NSData
 @return Model数组
 */
+ (nullable NSArray *)fwModelArrayWithJson:(id)json;

/**
 从json创建Model字典
 
 @param json json对象，支持NSDictionary、NSString、NSData
 @return Model字典
 */
+ (nullable NSDictionary *)fwModelDictionaryWithJson:(id)json;

/**
 从json对象设置对象属性
 
 @param json json对象，支持NSDictionary、NSString、NSData
 @return 是否设置成功
 */
- (BOOL)fwModelSetWithJson:(id)json;

/**
 从字典设置对象属性
 
 @param dictionary 字典数据
 @return 是否设置成功
 */
- (BOOL)fwModelSetWithDictionary:(NSDictionary *)dictionary;

/**
 转换为json对象
 
 @return json对象，如NSDictionary、NSArray，失败为nil
 */
- (nullable id)fwModelToJsonObject;

/**
 转换为json字符串数据
 
 @return NSData，失败为nil
 */
- (nullable NSData *)fwModelToJsonData;

/**
 转换为json字符串
 
 @return NSString，失败为nil
 */
- (nullable NSString *)fwModelToJsonString;

/**
 从属性拷贝当前对象
 
 @return 拷贝对象，失败为nil
 */
- (nullable id)fwModelCopy;

/// 对象编码
- (void)fwModelEncodeWithCoder:(NSCoder *)aCoder;

/// 对象解码
- (id)fwModelInitWithCoder:(NSCoder *)aDecoder;

/// 对象的hash编码
- (NSUInteger)fwModelHash;

/// 比较Model
- (BOOL)fwModelIsEqual:(id)model;

/// 对象描述
- (NSString *)fwModelDescription;

@end

/// Type-Encoding枚举
typedef NS_OPTIONS(NSUInteger, FWEncodingType) {
    FWEncodingTypeMask       = 0xFF, ///< mask of type value
    FWEncodingTypeUnknown    = 0, ///< unknown
    FWEncodingTypeVoid       = 1, ///< void
    FWEncodingTypeBool       = 2, ///< bool
    FWEncodingTypeInt8       = 3, ///< char / BOOL
    FWEncodingTypeUInt8      = 4, ///< unsigned char
    FWEncodingTypeInt16      = 5, ///< short
    FWEncodingTypeUInt16     = 6, ///< unsigned short
    FWEncodingTypeInt32      = 7, ///< int
    FWEncodingTypeUInt32     = 8, ///< unsigned int
    FWEncodingTypeInt64      = 9, ///< long long
    FWEncodingTypeUInt64     = 10, ///< unsigned long long
    FWEncodingTypeFloat      = 11, ///< float
    FWEncodingTypeDouble     = 12, ///< double
    FWEncodingTypeLongDouble = 13, ///< long double
    FWEncodingTypeObject     = 14, ///< id
    FWEncodingTypeClass      = 15, ///< Class
    FWEncodingTypeSEL        = 16, ///< SEL
    FWEncodingTypeBlock      = 17, ///< block
    FWEncodingTypePointer    = 18, ///< void*
    FWEncodingTypeStruct     = 19, ///< struct
    FWEncodingTypeUnion      = 20, ///< union
    FWEncodingTypeCString    = 21, ///< char*
    FWEncodingTypeCArray     = 22, ///< char[10] (for example)
    
    FWEncodingTypeQualifierMask   = 0xFF00,   ///< mask of qualifier
    FWEncodingTypeQualifierConst  = 1 << 8,  ///< const
    FWEncodingTypeQualifierIn     = 1 << 9,  ///< in
    FWEncodingTypeQualifierInout  = 1 << 10, ///< inout
    FWEncodingTypeQualifierOut    = 1 << 11, ///< out
    FWEncodingTypeQualifierBycopy = 1 << 12, ///< bycopy
    FWEncodingTypeQualifierByref  = 1 << 13, ///< byref
    FWEncodingTypeQualifierOneway = 1 << 14, ///< oneway
    
    FWEncodingTypePropertyMask         = 0xFF0000, ///< mask of property
    FWEncodingTypePropertyReadonly     = 1 << 16, ///< readonly
    FWEncodingTypePropertyCopy         = 1 << 17, ///< copy
    FWEncodingTypePropertyRetain       = 1 << 18, ///< retain
    FWEncodingTypePropertyNonatomic    = 1 << 19, ///< nonatomic
    FWEncodingTypePropertyWeak         = 1 << 20, ///< weak
    FWEncodingTypePropertyCustomGetter = 1 << 21, ///< getter=
    FWEncodingTypePropertyCustomSetter = 1 << 22, ///< setter=
    FWEncodingTypePropertyDynamic      = 1 << 23, ///< @dynamic
};

/// 解析Type-Encoding字符串类型
FWEncodingType FWEncodingGetType(const char *typeEncoding);

/// Ivar信息
@interface FWClassIvarInfo : NSObject

@property (nonatomic, assign, readonly) Ivar ivar;              ///< ivar opaque struct
@property (nonatomic, strong, readonly) NSString *name;         ///< Ivar's name
@property (nonatomic, assign, readonly) ptrdiff_t offset;       ///< Ivar's offset
@property (nonatomic, strong, readonly) NSString *typeEncoding; ///< Ivar's type encoding
@property (nonatomic, assign, readonly) FWEncodingType type;    ///< Ivar's type

- (instancetype)initWithIvar:(Ivar)ivar;

@end

/// Method 信息
@interface FWClassMethodInfo : NSObject

@property (nonatomic, assign, readonly) Method method;                  ///< method opaque struct
@property (nonatomic, strong, readonly) NSString *name;                 ///< method name
@property (nonatomic, assign, readonly) SEL sel;                        ///< method's selector
@property (nonatomic, assign, readonly) IMP imp;                        ///< method's implementation
@property (nonatomic, strong, readonly) NSString *typeEncoding;         ///< method's parameter and return types
@property (nonatomic, strong, readonly) NSString *returnTypeEncoding;   ///< return value's type
@property (nullable, nonatomic, strong, readonly) NSArray<NSString *> *argumentTypeEncodings; ///< array of arguments' type

- (instancetype)initWithMethod:(Method)method;

@end

/// 属性信息
@interface FWClassPropertyInfo : NSObject

@property (nonatomic, assign, readonly) objc_property_t property; ///< property's opaque struct
@property (nonatomic, strong, readonly) NSString *name;           ///< property's name
@property (nonatomic, assign, readonly) FWEncodingType type;      ///< property's type
@property (nonatomic, strong, readonly) NSString *typeEncoding;   ///< property's encoding value
@property (nonatomic, strong, readonly) NSString *ivarName;       ///< property's ivar name
@property (nullable, nonatomic, assign, readonly) Class cls;      ///< may be nil
@property (nullable, nonatomic, strong, readonly) NSArray<NSString *> *protocols; ///< may nil
@property (nonatomic, assign, readonly) SEL getter;               ///< getter (nonnull)
@property (nonatomic, assign, readonly) SEL setter;               ///< setter (nonnull)

- (instancetype)initWithProperty:(objc_property_t)property;

@end

/// Class信息
@interface FWClassInfo : NSObject

@property (nonatomic, assign, readonly) Class cls; ///< class object
@property (nullable, nonatomic, assign, readonly) Class superCls; ///< super class object
@property (nullable, nonatomic, assign, readonly) Class metaCls;  ///< class's meta class object
@property (nonatomic, readonly) BOOL isMeta; ///< whether this class is meta class
@property (nonatomic, strong, readonly) NSString *name; ///< class name
@property (nullable, nonatomic, strong, readonly) FWClassInfo *superClassInfo; ///< super class's class info
@property (nullable, nonatomic, strong, readonly) NSDictionary<NSString *, FWClassIvarInfo *> *ivarInfos; ///< ivars
@property (nullable, nonatomic, strong, readonly) NSDictionary<NSString *, FWClassMethodInfo *> *methodInfos; ///< methods
@property (nullable, nonatomic, strong, readonly) NSDictionary<NSString *, FWClassPropertyInfo *> *propertyInfos; ///< properties

- (void)setNeedUpdate;

- (BOOL)needUpdate;

+ (nullable instancetype)classInfoWithClass:(Class)cls;

+ (nullable instancetype)classInfoWithClassName:(NSString *)className;

@end

NS_ASSUME_NONNULL_END
